Adds Power Spectrum example for a timeseries

yashkgp opened the pull request:
Issue #2483 
Here is an example : 
```python
import sunpy.timeseries
import sunpy.data.sample
import matplotlib.pyplot as plt
import pandas as pd 
from scipy import signal
ts = sunpy.timeseries.TimeSeries(sunpy.data.sample.GOES_XRS_TIMESERIES)
k,l=ts.power_spectra()
plt.semilogy(k[0], l[0])
plt.show()
```
![image](https://user-images.githubusercontent.com/24179111/36917027-a726d6b4-1e7b-11e8-8c99-23cb4e7699bd.png)


Apart from adding the unit tests,is anything else needed for this method?

pep8speaks commented:
Hello @yashkgp! Thanks for submitting the PR.

 - In the file [`sunpy/timeseries/timeseriesbase.py`](https://github.com/sunpy/sunpy/blob/a488175c006ed957a0b014d9e15e5f20a432a557/sunpy/timeseries/timeseriesbase.py), following are the PEP8 issues :

> [Line 137:9](https://github.com/sunpy/sunpy/blob/a488175c006ed957a0b014d9e15e5f20a432a557/sunpy/timeseries/timeseriesbase.py#L137): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 138:9](https://github.com/sunpy/sunpy/blob/a488175c006ed957a0b014d9e15e5f20a432a557/sunpy/timeseries/timeseriesbase.py#L138): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 179:13](https://github.com/sunpy/sunpy/blob/a488175c006ed957a0b014d9e15e5f20a432a557/sunpy/timeseries/timeseriesbase.py#L179): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 211:13](https://github.com/sunpy/sunpy/blob/a488175c006ed957a0b014d9e15e5f20a432a557/sunpy/timeseries/timeseriesbase.py#L211): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 212:13](https://github.com/sunpy/sunpy/blob/a488175c006ed957a0b014d9e15e5f20a432a557/sunpy/timeseries/timeseriesbase.py#L212): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 241:101](https://github.com/sunpy/sunpy/blob/a488175c006ed957a0b014d9e15e5f20a432a557/sunpy/timeseries/timeseriesbase.py#L241): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (138 > 100 characters)
> [Line 272:16](https://github.com/sunpy/sunpy/blob/a488175c006ed957a0b014d9e15e5f20a432a557/sunpy/timeseries/timeseriesbase.py#L272): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 276:16](https://github.com/sunpy/sunpy/blob/a488175c006ed957a0b014d9e15e5f20a432a557/sunpy/timeseries/timeseriesbase.py#L276): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 313:101](https://github.com/sunpy/sunpy/blob/a488175c006ed957a0b014d9e15e5f20a432a557/sunpy/timeseries/timeseriesbase.py#L313): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 100 characters)
> [Line 319:101](https://github.com/sunpy/sunpy/blob/a488175c006ed957a0b014d9e15e5f20a432a557/sunpy/timeseries/timeseriesbase.py#L319): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (127 > 100 characters)
> [Line 430:101](https://github.com/sunpy/sunpy/blob/a488175c006ed957a0b014d9e15e5f20a432a557/sunpy/timeseries/timeseriesbase.py#L430): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (113 > 100 characters)
> [Line 503:33](https://github.com/sunpy/sunpy/blob/a488175c006ed957a0b014d9e15e5f20a432a557/sunpy/timeseries/timeseriesbase.py#L503): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 528:31](https://github.com/sunpy/sunpy/blob/a488175c006ed957a0b014d9e15e5f20a432a557/sunpy/timeseries/timeseriesbase.py#L528): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 528:33](https://github.com/sunpy/sunpy/blob/a488175c006ed957a0b014d9e15e5f20a432a557/sunpy/timeseries/timeseriesbase.py#L528): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 549:32](https://github.com/sunpy/sunpy/blob/a488175c006ed957a0b014d9e15e5f20a432a557/sunpy/timeseries/timeseriesbase.py#L549): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'
> [Line 555:21](https://github.com/sunpy/sunpy/blob/a488175c006ed957a0b014d9e15e5f20a432a557/sunpy/timeseries/timeseriesbase.py#L555): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 629:101](https://github.com/sunpy/sunpy/blob/a488175c006ed957a0b014d9e15e5f20a432a557/sunpy/timeseries/timeseriesbase.py#L629): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 100 characters)

dpshelio commented on sunpy/timeseries/timeseriesbase.py:
```diff
@@ -524,6 +525,35 @@ def _sanitize_metadata(self, **kwargs):
+            data_ = self.data[col]
+            freq, spectra = signal.periodogram(data_, fs)
+            Freqs.append(freq)
+            Spectra.append(spectra)
```
```python
from functools import partial
periodogram_fs = partial(signal.periodogram, fs=fs)
Freqs, Spectra = zip(*map(periodogram_fs, self.data))
```
This will remove all the loop, it will produce tuples instead of lists though.

dpshelio commented on sunpy/timeseries/timeseriesbase.py:
```diff
@@ -524,6 +525,35 @@ def _sanitize_metadata(self, **kwargs):
+
+        Returns
+        -------
+        Freqs : array of arrays
```
You are not returning `array` but a `list` of arrays.

dpshelio commented on sunpy/timeseries/timeseriesbase.py:
```diff
@@ -524,6 +525,35 @@ def _sanitize_metadata(self, **kwargs):
+
+        Parameter
+        ----------
+        fs : `float, optional`
```
optional should be outside the backtick.

nabobalis commented:
Hey @yashkgp, sorry for the late reply.

We aren't sure we want to add methods on to time series right now. It sparked quite a discussion about methods in various parts of SunPy. That shake the foundation of SunPy itself. 

What we want from this now is if you could turn it into an example we could add to our [Gallery](http://docs.sunpy.org/en/stable/generated/gallery/index.html)? If you have a [look at the other examples here](https://github.com/sunpy/sunpy/tree/master/examples) for how we format them!

yashkgp commented:
Sure @nabobalis Not a Problem .

yashkgp commented:
Hey @nabobalis Could you give a review.

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+"""
+===============================
+POWER SPECTRUM OF A TIMESERIES
```
Not all caps, just normal case.

yashkgp replied:
Done.

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+POWER SPECTRUM OF A TIMESERIES
+===============================
+
+An example showing how to estimate power spectrum of a TimeSeries
```
I think a `the` should between `estimate` and `power`. Full stop at the end.

yashkgp replied:
Done.

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+###############################################################################
+# Let's first load a TimeSeries from sample data
+
+ts = sunpy.timeseries.TimeSeries(GOES_XRS_TIMESERIES)
```
You might want to describe the data you will working on to give whoever reads this an idea of what the input is.

yashkgp replied:
I have added a brief description about the input data now  . Is it okay ?

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+ts = sunpy.timeseries.TimeSeries(GOES_XRS_TIMESERIES)
+
+###############################################################################
+# We now use scipy's periodogram to estimate the power spectra of the first 
```
You want to explain why you use the periodogram instead of say the FFT. Also we probably want to mention the limitations or requirements  of using these methods. We would like people not to assume they can use these methods on any signal.  

I think there is one in Astropy as well now. But I am unsure of what version.

yashkgp replied:
Yeah , just discovered that . Astropy's Periodogram has some amazing functionalities . We should probably switch over to Astropy's Periodogram .

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+# column of the Timeseries
+
+cols=ts.columns
+freq, spectra = signal.periodogram(ts.data[cols[0]], fs=1)
```
Is `fs=1` correct for the data? 

Also I wonder if we could use the false alarm probability in this example as well.

yashkgp replied:
Yeah , this feature is covered in Astropy's periodogram . Should we switch over to Astropy's periodogram ?

nabobalis replied:
If it works out regarding the version of Astropy it is supported in.

yashkgp replied:
Yeah , it's there in Astropy 3.0

nabobalis replied:
Well we still have to support Astropy 2 until our next release.

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+
+
+
+
```
All these extra lines need to go.

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+
+plt.semilogy(freq, spectra)
+plt.title('Power Spectrum of {}'.format(cols[0]))
+plt.ylabel('Power Spectral Density')
```
Missing the unit

yashkgp commented:
Hey @nabobalis , Just discovered that Astropy's Periodogram is much more powerful than scipy's Periodogram . They have full fledged tutorial dedicated to Periodogram [http://docs.astropy.org/en/stable/stats/lombscargle.html#](url).It has various functionalities like frequency grid spacing, false alarm probabilities, normalization etc.We should definitely switch over to Astrpoy's peiodogram . What do you say @nabobalis ?

dpshelio commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+# We now use scipy's periodogram to estimate the power spectra of the first 
+# column of the Timeseries
+
+cols=ts.columns
```
Since probably these examples need to be somehow meaningful I would do the following instead of using a variable called cols:
```python
xrays_short = ts.columns[0]
```
and change `cols[0]` for `xrays_short` everywhere. That makes the code more readable.

yashkgp replied:
Done .

yashkgp commented:
Hey @nabobalis @dpshelio , I tried switching to Astropy's version of periodogram (yeah, its supported in Astropy 2 ) but its throwing a ValueError
`TypeError: ufunc true_divide cannot use operands with types dtype('float64') and dtype('<m8[ns]')`
It seems like Astropy's periodogram does accept Pandas Timeseries object . One possible solution might to use Timedelta and convert `dtype('<m8[ns]')` into float . Is there any easy way to solve this problem ?

nabobalis commented:
If you can't get the astropy one to work, It would just be worth noting that its available and linking to it in the example.

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,41 @@
+ts = sunpy.timeseries.TimeSeries(GOES_XRS_TIMESERIES)
+# Data is taken from GOES X-Ray Sensor(XRS), which observes full-disk-integrated
+# solar flux in two broadband channels: 1--8 angstrom (long); and
+# 0.5--4 angstrom (short)
```
This should be moved above the python code. 

But what I wanted here was not what is GOES, it was can we use signal analysis on this data. Is it evenly spaced in time, whats its time step, are there any problems that might mean that signal analysis would be tricky etc. We don't want to give the impression you can just singal analysis any signal.

yashkgp replied:
Data is not completely evenly sampled,it has 2 outliers .Should I mention this?

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,41 @@
+
+###############################################################################
+# We now use scipy's periodogram to estimate the power spectra of the first 
+# column of the Timeseries
```
Make a link to the astropy version here, saying it's available instead.

dpshelio commented on examples/power_spectra_example.py:
```diff
@@ -17,23 +17,26 @@
-cols=ts.columns
-freq, spectra = signal.periodogram(ts.data[cols[0]], fs=1)
+xray_short = ts.columns[0]
+freq, spectra = signal.periodogram(ts.data[xray_short], fs=1)
```
I've looked in detail into the data and I don't believe this is correct. `fs=1` means the cadence of the data in Hz, the data seems to not be completely evenly sampled!!
```python
>>> sample = np.array([(x - y).total_seconds() for x, y in zip(ts.index[1:], ts.index)])
>>> sample.min(), sample.max(), sample.mean(), sample.std()
  (2.0430000000000001, 14.337000000000002, 2.0485031771623676, 0.077912702228747382)
```
I don't know how there's a 14.33 seconds gap!!! If you plot that you see it happens twice! This means that this analysis would be wrong all together!

In case the cadence would be 2.043, then I think the `fs` should be `1/2.043 s = 0.48 Hz`

yashkgp replied:
Should I clean the data by removing those 2 outliers ?

yashkgp replied:
Should I just change `fs` to 0.48 and notify the users about the irregularities in data or should I change the input data ?

dpshelio commented on examples/power_spectra_example.py:
```diff
@@ -17,23 +17,26 @@
-plt.title('Power Spectrum of {}'.format(cols[0]))
-plt.ylabel('Power Spectral Density')
+plt.title('Power Spectrum of {}'.format(xray_short))
+plt.ylabel('Power Spectral Density(({})**2/Hz)'.format(ts.units[xray_short]))
```
You could use units here as:
```python
import astropy.units as u
plt.ylabel('Power Spectral Density ({:LaTeX})'.format(ts.units[xray_short] ** 2 / u.Hz))
```
Also, put a space between the `Density` and the `()`

dpshelio replied:
I believe square brackets for the units are more common: `[Hz]` instead of `(Hz)`

yashkgp replied:
Yeah, this is much cleaner.

dpshelio commented on examples/power_spectra_example.py:
```diff
@@ -17,23 +17,26 @@
-plt.ylabel('Power Spectral Density')
+plt.title('Power Spectrum of {}'.format(xray_short))
+plt.ylabel('Power Spectral Density(({})**2/Hz)'.format(ts.units[xray_short]))
 plt.xlabel('Frequency(Hz)')
```
Also space between the title and the units.

yashkgp commented:
Hey @nabobalis @dpshelio , I have changed the input data, as the previous one was not evenly sampled . I have switched to `RHESSI_TIMESERIES` , which is a evenly sampled timeseries with a time step of 4 sec. Could you please give it a review?

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+import matplotlib.pyplot as plt
+import pandas as pd 
+from scipy import signal
+
```
I would remove this blank line.

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+
+###############################################################################
+# Let's first load a TimeSeries from sample data.
+
```
Remove this line.

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+ts = sunpy.timeseries.TimeSeries(RHESSI_TIMESERIES)
+
+###############################################################################
+# We now use scipy's periodogram to estimate the power spectra of the first 
```
SciPy, also what is the first column?

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+###############################################################################
+# We now use scipy's periodogram to estimate the power spectra of the first 
+# column of the Timeseries. Astropy's Lomb-Scargle Periodograms can be used
+# instead of scipy's periodogram.
```
SciPy

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+
+###############################################################################
+# We now use scipy's periodogram to estimate the power spectra of the first 
+# column of the Timeseries. Astropy's Lomb-Scargle Periodograms can be used
```
Link to Astropy's documentation for it.

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+x_ray = ts.columns[0]
+# The suitable value for fs would be 0.25 Hz as the time step is 4 s.
+freq, spectra = signal.periodogram(ts.data[xray_short], fs=0.25)
+###############################################################################
```
Blank line here above the `####...`

yashkgp replied:
Done

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+
+x_ray = ts.columns[0]
+# The suitable value for fs would be 0.25 Hz as the time step is 4 s.
+freq, spectra = signal.periodogram(ts.data[xray_short], fs=0.25)
```
x_ray

yashkgp replied:
Sorry @nabobalis ,had updated in my local branch , but somehow forgot to push it.

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+# Plot the power spectrum
+
+plt.semilogy(freq, spectra)
+plt.title('Power Spectrum of {}'.format(xray_short))
```
x_ray

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,44 @@
+# Let's first load a TimeSeries from sample data.
+# Input data contains 9 columns, which are evenly sampled with a time
+# step of 4 seconds. So, the data is suitable for signal analysis.
+ts = sunpy.timeseries.TimeSeries(RHESSI_TIMESERIES)
```
Line break above here just to be consistent.

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,44 @@
+from sunpy.data.sample import RHESSI_TIMESERIES
+
+###############################################################################
+# Let's first load a TimeSeries from sample data.
```
`Lets first load a RHESSI TimeSeries from SunPy's sample data`

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,44 @@
+
+###############################################################################
+# Let's first load a TimeSeries from sample data.
+# Input data contains 9 columns, which are evenly sampled with a time
```
`This data contains`

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,44 @@
+###############################################################################
+# Let's first load a TimeSeries from sample data.
+# Input data contains 9 columns, which are evenly sampled with a time
+# step of 4 seconds. So, the data is suitable for signal analysis.
```
Remove the last sentence.

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,44 @@
+ts = sunpy.timeseries.TimeSeries(RHESSI_TIMESERIES)
+
+###############################################################################
+# We now use SciPy's periodogram to estimate the power spectra of the first 
```
I think that `~scipy.signal.lombscargle` should work as a link to the documentation

yashkgp replied:
Ohh! Thats quite useful. Thanks a lot @nabobalis

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,44 @@
+# We now use SciPy's periodogram to estimate the power spectra of the first 
+# column of the Timeseries. The first column contains X-Ray emmisions in the range 
+# of 3-6 keV. Astropy's Lomb-Scargle Periodograms 
+# (http://docs.astropy.org/en/v2.0.5/stats/lombscargle.html) can be used
```
Try `~astropy.stats.LombScargle` instead. Sphinx should link to it since we have that setup. 

I would rephrase the last sentence as well. `An alterantive version is ....` or something.

dpshelio commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,44 @@
+# Start by importing the necessary modules.
+
+import matplotlib.pyplot as plt
+import pandas as pd 
```
pandas is not used

dpshelio commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,44 @@
+ts = sunpy.timeseries.TimeSeries(RHESSI_TIMESERIES)
+
+###############################################################################
+# We now use SciPy's periodogram(~scipy.signal.lombscargle) to estimate the power 
```
Wait... `signal.periodogram` is the same than `signal.lombscargle`? If so, then you could have used the previous series that was not evenly spaced.

yashkgp replied:
@dpshelio Its not the same , I'll just change it in the next commit .
Sorry @dpshelio , I am not that acquainted with different types of periodogram. Should I use a normal `signal.periodogram` or a `signal.lombscargle`?

yashkgp replied:
Sorry @dpshelio , I am not that well acquainted with lomb scargle periodogram . Is there any advantage of using `signal.lombscargle` over `signal.periodogram`?

nabobalis replied:
They are not the same, that should point to `scipy.signal.periodogram` instead.

nabobalis commented:
Odd how the sphinx interlink hasn't worked. I wonder what is wrong.

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+ts = sunpy.timeseries.TimeSeries(RHESSI_TIMESERIES)
+
+###############################################################################
+# We now use SciPy's periodogram(~scipy.signal.periodogram) to estimate the power 
```
There needs to be ` on each side for these links to work.

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+# We now use SciPy's periodogram(~scipy.signal.periodogram) to estimate the power 
+# spectra of the first column of the Timeseries. The first column contains X-Ray 
+# emmisions in the range of 3-6 keV. An alterantive version is Astropy's 
+# Lomb-Scargle Periodograms(~astropy.stats.LombScargle).
```
There needs to be ` on each side for these links to work.

yashkgp commented:
Sorry for the late update , had my exams going on.

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+ts = sunpy.timeseries.TimeSeries(RHESSI_TIMESERIES)
+
+###############################################################################
+# We now use SciPy's periodogram(`~scipy.signal.periodogram`) to estimate the 
```
Looking at the final result, i think we want this to be `SciPy's ~scipy.signal.periodogram`

nabobalis commented on examples/power_spectra_example.py:
```diff
@@ -0,0 +1,43 @@
+# We now use SciPy's periodogram(`~scipy.signal.periodogram`) to estimate the 
+# power spectra of the first column of the Timeseries. The first column contains
+# X-Ray emmisions in the range of 3-6 keV. An alterantive version is Astropy's 
+# Lomb-Scargle Periodograms(`~astropy.stats.LombScargle`).
```
Similar here. `Astropy's  ~astropy.stats.LombScargle`.

nabobalis commented:
Thanks @yashkgp!
